Mathlib/Analysis/LocallyConvex/Barrelled.lean

```diff
@@ -0,0 +1,193 @@
+  refine Seminorm.continuous_of_lowerSemicontinuous _ ?_
+  rw [Seminorm.coe_iSup_eq bdd]
+  rw [Seminorm.bddAbove_range_iff] at bdd
+  convert lowerSemicontinuous_ciSup (f := fun i x ↦ p i x) bdd (fun i ↦ (hp i).lowerSemicontinuous)
+  exact iSup_apply
+
+end defs
+
+section TVS_anyField
+
+variable {α ι κ 𝕜₁ 𝕜₂ E F : Type*} [Nonempty κ] [NontriviallyNormedField 𝕜₁]
+    [NontriviallyNormedField 𝕜₂] {σ₁₂ : 𝕜₁ →+* 𝕜₂} [RingHomIsometric σ₁₂]
+    [AddCommGroup E] [AddCommGroup F] [Module 𝕜₁ E] [Module 𝕜₂ F]
+
+/-- Any TVS over a `NontriviallyNormedField` that is also a Baire space is barrelled. In
+particular, this applies to Banach spaces and Fréchet spaces. -/
+instance BaireSpace.instBarrelledSpace [TopologicalSpace E] [TopologicalAddGroup E]
+    [ContinuousConstSMul 𝕜₁ E] [BaireSpace E] :
+    BarrelledSpace 𝕜₁ E where
+  continuous_of_lowerSemicontinuous := by
+    -- Let `p` be a lower-semicontinuous seminorm on `E`.
+    intro p hp
+    -- Consider the family of all `p`-closed-balls with integer radius.
+    -- By lower semicontinuity, each of these closed balls is indeed closed...
+    have h₁ : ∀ n : ℕ, IsClosed (p.closedBall (0 : E) n) := fun n ↦ by
+      simpa [p.closedBall_zero_eq] using hp.isClosed_preimage n
+    -- ... and clearly they cover the whole space.
+    have h₂ : (⋃ n : ℕ, p.closedBall (0 : E) n) = univ :=
+      eq_univ_of_forall fun x ↦ mem_iUnion.mpr (exists_nat_ge <| p (x - 0))
+    -- Hence, one of them has nonempty interior. Let `n : ℕ` be its radius, and fix `x` an
+    -- interior point.
+    rcases nonempty_interior_of_iUnion_of_closed h₁ h₂ with ⟨n, ⟨x, hxn⟩⟩
+    -- To show that `p` is continuous, we will show that the `p`-closed-ball of
+    -- radius `2*n` is a neighborhood of zero.
+    refine Seminorm.continuous' (r := 2*n) ?_
+    rw [p.closedBall_zero_eq] at hxn ⊢
+    have hxn' : p x ≤ n := by convert interior_subset hxn
+    -- By definition, we have `p x' ≤ n` for `x'` sufficiently close to `x`.
+    -- In other words, `p (x + y) ≤ n` for `y` sufficiently close to `0`.
+    rw [mem_interior_iff_mem_nhds, ← map_add_left_nhds_zero] at hxn
+    -- Hence, for `y` sufficiently close to `0`, we have
+    -- `p y = p (x + y - x) ≤ p (x + y) + p x ≤ 2*n`
+    filter_upwards [hxn] with y hy
+    calc p y = p (x + y - x) := by rw [add_sub_cancel']
+      _ ≤ p (x + y) + p x := map_sub_le_add _ _ _
+      _ ≤ n + n := add_le_add hy hxn'
+      _ = 2*n := by ring
+
+namespace WithSeminorms
+
+variable [UniformSpace E] [UniformSpace F] [UniformAddGroup E] [UniformAddGroup F]
+    [ContinuousSMul 𝕜₁ E] [ContinuousSMul 𝕜₂ F] [BarrelledSpace 𝕜₁ E] {𝓕 : ι → E →SL[σ₁₂] F}
+    {q : SeminormFamily 𝕜₂ F κ} (hq : WithSeminorms q)
+
+/-- The **Banach-Steinhaus** theorem for maps from a barrelled space to any space whose topology is
+generated by a family of seminorms. Use `WithSeminorms.equicontinuous_TFAE` and
+`Seminorm.bound_of_continuous` to get explicit bounds on the seminorms from equicontinuity. -/
```

Comment:
It would also be nice to mention the term *Uniform Boundedness Principle* here to aid in searching the docs.